src/types/web-resource.ts

```diff
@@ -0,0 +1,108 @@
+import * as TypeUtils from '../type-utils';
+import { storageRegistry } from '../storage-adapters/index';
+
+/**
+ * This is how WebResources are stored on the DB
+ */
+export type WebResourceRef = {
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+/**
+ * This is how WebResources are defined on creation
+ */
+export type WebResourceInput = {
+	filename: string;
+	data: Buffer;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+	storage: string;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
+	},
+};
+
+export const fetchProcessing = (data: any) => {
+	let refData: WebResourceRef;
+	if (data === null) {
+		return data;
+	}
+	if (typeof data === 'string') {
+		refData = JSON.parse(data);
+	} else if (typeof data === 'object') {
+		refData = data;
+	} else {
+		throw new Error(
+			'Unable to parse data for a WebResource type. Data is: ' + data,
+		);
+	}
+
+	return {
+		filename: refData.filename,
+		href: refData.href,
+		contentType: refData.contentType,
+		contentDisposition: refData.contentDisposition,
+		size: refData.size,
+	};
+};
+
+export const validate = TypeUtils.validate.checkRequired(
+	async (value: WebResourceInput) => {
+		if (typeof value !== 'object') {
+			throw new Error('received value is not an object');
+		}
+		if (!value.filename) {
+			throw new Error('filename is required');
+		}
+		if (!value.data) {
+			throw new Error('data is required');
+		}
+
+		const storage = value.storage;
+		if (!storage) {
+			throw new Error('storage is required');
+		}
+		const storageAdapter = storageRegistry[storage];
+		if (!storageAdapter) {
+			throw new Error(
+				`storage ${storage} not defined. Available adapters: ${Object.keys(
+					storageRegistry,
+				)}`,
+			);
```

Comment:
This is awkward and very much not an error I'd want to expose to an external API request, I think ideally this should be specified/checked whilst setting up the field unless there is some strong benefit of being able to change the storage for individual requests?

Reply:
The rationale behind this is that the `storage` field is sent by the client app, so we need to validate it to be valid at runtime, since from `sbvr-types` we don't know which adapters are present.

Being able to specify the `storage` to use on each call gives the app flexibility to configure several adapters if needed, for example, if it needs/wants to use different S3 buckets, or storage classes, or even other adapters. 

If by `whilst setting up the field` you mean when defining the property of this type, then I think it could be done but I'm not sure how that is defined. I haven't seen an example of setting a default value for a term property.